contracts/programs/data-feeds-cache/src/lib.rs

```diff
@@ -0,0 +1,1094 @@
+
+declare_id!("3kX63udXtYcsdj2737Wi2KGd2PhqiKPgAFAxstrjtRUa");
+
+mod common;
+mod context;
+mod error;
+mod event;
+pub mod state;
+
+use common::{ANCHOR_DISCRIMINATOR, MAX_WORKFLOW_METADATAS, ZERO_ADDRESS, ZERO_DATA_ID};
+use context::*;
+use error::{AuthError, DataCacheError};
+use event::{DecimalFeedConfigSet, LegacyFeedsReported};
+use state::{
+    AdminList, CacheTransmission, DecimalReport, FeedConfig, LegacyFeedEntry, LegacyFeedsConfig,
+    ReceivedDecimalReport, WorkflowMetadata, WritePermissionFlag,
+};
+
+use anchor_lang::solana_program::hash;
+#[program]
+pub mod data_feeds_cache {
+
+    use anchor_lang::{
+        solana_program::{instruction::Instruction, program::invoke_signed, system_instruction},
+        Discriminator,
+    };
+
+    use super::*;
+
+    // todo, add feed admins? who will also be a zero_copy array
+    pub fn initialize(ctx: Context<Initialize>, feed_admins: Vec<Pubkey>) -> Result<()> {
+        let state = &mut ctx.accounts.state.load_init()?;
+        state.owner = ctx.accounts.owner.key();
+
+        let mut prev_admin = Pubkey::default();
+        for admin in feed_admins.iter() {
+            require!(
+                &prev_admin <= admin,
+                DataCacheError::AddressesMustStrictlyIncrease
+            );
+            state.feed_admins.push(*admin);
+            prev_admin = *admin;
+        }
+
+        // todo: can probably remove
+        state.proposed_owner = Pubkey::default();
+
+        let (_, authority_nonce) = Pubkey::find_program_address(
+            &[b"legacy_writer", ctx.accounts.state.key().as_ref()],
+            &crate::ID,
+        );
+        state.legacy_writer_nonce = authority_nonce;
+
+        Ok(())
+    }
+
+    pub fn set_feed_admin(ctx: Context<SetFeedAdmin>, admin: Pubkey, is_admin: bool) -> Result<()> {
+        let mut state = ctx.accounts.state.load_mut()?;
+
+        match state.feed_admins.binary_search_by(|a| a.cmp(&admin)) {
+            Ok(i) => {
+                if !is_admin {
+                    state.feed_admins.remove(i);
+                }
+            }
+            Err(i) => {
+                if is_admin {
+                    state.feed_admins.insert(i, admin);
+                }
+            }
+        };
+        Ok(())
+    }
+
+    pub fn transfer_ownership(
+        ctx: Context<TransferOwnership>,
+        proposed_owner: Pubkey,
+    ) -> Result<()> {
+        let state = &mut ctx.accounts.state.load_mut()?;
+        state.proposed_owner = proposed_owner;
+
+        Ok(())
+    }
+
+    pub fn accept_ownership(ctx: Context<AcceptOwnership>) -> Result<()> {
+        let state = &mut ctx.accounts.state.load_mut()?;
+
+        state.owner = state.proposed_owner;
+        state.proposed_owner = Pubkey::default();
+
+        Ok(())
+    }
+
+    pub fn close_decimal_reports<'info>(
+        ctx: Context<CloseDecimalReports<'info>>,
+        data_ids: Vec<[u8; 16]>,
+    ) -> Result<()> {
+        let state = &mut ctx.accounts.state.load()?;
```

Comment:
It seems this doesn't have to be `mut` here? 
```suggestion
        let state = &ctx.accounts.state.load()?;
```